bftengine/include/bftengine/ControlStateManager.hpp

```diff
@@ -0,0 +1,42 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+#include <string>
+#include <unordered_map>
+
+namespace bftEngine {
+
+class IControlStateManager {
+ public:
+  enum ControlStateId {
```

Comment:
If we make this a [std::variant](https://en.cppreference.com/w/cpp/utility/variant) we don't need separate `ControlStateId`s and values, and we wouldn't need to tie our value type to a string, which likely doesn't make much sense for many uses of this. We could use the [index of the variant](https://en.cppreference.com/w/cpp/utility/variant/index) as the id for serialization into reserved pages.

An example is something like this:

```c++
struct StopAtNextCheckpoint {};
struct RotateKeys{
    std::map<..> newKeys;
}

using ControlStateOperation = std::variant<StopAtNextCheckpoint, RotateKeys>;

using CurrentCheckpointForUpgrade = std::optional<uint64_t>;
using CurrentKeys = std::map<..>;

using ControlState = std::variant<CurrentCheckpointForUpgrade, CurrentKeys>;

// Throws on Error
void setControlState(ControlStateOperation op); 

// We use the index of the variant for retrieval. 
ControlState getControlState(std::size_t index);
```

While the above matches some future plans I have for message based systems, it seems unnecessarily complex for now. Equally unnecessary, would be to use virtual types for operations and states, where each instantiation of `IControlStateOperation` or `IControlState` has a unique ID and can be registered.

For now, since we are pretty limited in what we are using this for, and will likely evolve it in the future, why not just do the simplest possible thing and provide a concrete type with functions for each state? The internal implementation can map the functions to ids for serialization. Something like this:

```
class ControlStateManager {
  public:
   void stopAtNextCheckpoint();
   std::optional<uint64_t> getUpgradeCheckpoint();
   void SetKeys(std::map<..> Keys);
   std::map<..> getKeys();
   
 private:
  ControlState some_state_structure_with_ids_that_lives_in_reserved_pages_;
  IStateTransfer& reserved_pages_api_;
};
```
If you really needed to, you could make this interface abstract, to allow you to use a test version, but that maybe overkill as well. Instead the single, concrete, `ControlStateManager` would take the reserved pages interface as an abstract interface, which we already have defined in [IStateTransfer](https://github.com/vmware/concord-bft/blob/master/bftengine/include/bftengine/IStateTransfer.hpp#L52-L57)

Reply:
I built a basic nonabstract implementation with a method for each controlState (as you suggested) based on the state transfer module.

For now the implementation is empty until @toly-kournik will finish his adaptions for the reserved pages and the state transfer module which will also answer our needs for allocating specific pages for the controlState data

Reply:
Thank you!